skribenten-backend/src/main/kotlin/no/nav/pensjon/brev/skribenten/services/BrevmalService.kt

```diff
@@ -0,0 +1,177 @@
+package no.nav.pensjon.brev.skribenten.services
+
+import io.ktor.server.application.*
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import no.nav.pensjon.brev.skribenten.services.BrevdataDto.BrevkategoriCode.*
+import no.nav.pensjon.brev.skribenten.services.BrevdataDto.BrevregeltypeCode
+import no.nav.pensjon.brev.skribenten.services.BrevdataDto.BrevregeltypeCode.*
+import no.nav.pensjon.brev.skribenten.services.PenService.SakType.*
+import org.slf4j.LoggerFactory
+
+class BrevmalService(
+    private val penService: PenService,
+    private val brevmetadataService: BrevmetadataService,
+) {
+    private val logger = LoggerFactory.getLogger(BrevmalService::class.java)
+
+    suspend fun hentBrevmalerForSak(
+        sakType: PenService.SakType,
+        includeEblanketter: Boolean,
+    ): List<LetterMetadata> = coroutineScope {
+        brevmetadataService.hentMaler(sakType, includeEblanketter).maler
+            .filter { filterForKontekst(it, isVedtaksKontekst = false) }
+            .filter { it.redigerbart }
+            .plus(brevmetadataService.hentMaler(sakType, includeEblanketter).eblanketter)
+            .filter { !utfiltrerteBrev.contains(it.brevkodeIBrevsystem) }
+            .map { it.mapToMetadata() }
+    }
+
+    suspend fun hentBrevmalerForVedtak(
+        sakType: PenService.SakType,
+        call: ApplicationCall,
+        includeEblanketter: Boolean,
+        vedtaksId: String,
+    ): List<LetterMetadata> = coroutineScope {
+        val brevmetadataAsync = async { brevmetadataService.hentMaler(sakType, includeEblanketter) }
+        val erKravPaaGammeltRegelverk =
+            async {
+                if (sakType == PenService.SakType.ALDER) {
+                    penService.hentIsKravPaaGammeltRegelverk(call, vedtaksId)
+                        .catch { message, httpStatusCode ->
+                            logger.error("Feil ved henting av felt \"erKravPaaGammeltRegelverk\" fra vedtak. Status: $httpStatusCode, message: $message")
+                            false
+                        }
+                } else false
+            }.await()
+
+        val relevanteMaler = brevmetadataAsync.await().maler
+            .filter { filterForKontekst(it, isVedtaksKontekst = true) }
+            .filter { it.redigerbart }
+            .filter {  erRelevantRegelverkstypeForSaktype(sakType, it.brevregeltype, erKravPaaGammeltRegelverk) }
+        return@coroutineScope relevanteMaler
+            .plus<BrevdataDto>(brevmetadataAsync.await().eblanketter)
+            .filter<BrevdataDto> { !utfiltrerteBrev.contains(it.brevkodeIBrevsystem) }
+            .map<BrevdataDto, LetterMetadata> { it.mapToMetadata() }
+
+    }
+
+    private fun filterForKontekst(brevmetadata: BrevdataDto, isVedtaksKontekst: Boolean): Boolean =
```

Comment:
Denne vil kanskje være litt lettere å lese delt opp i to funksjoner; `filterVedtakskontekst` og `filterSakskontekst`.

Reply:
Done